chain/blocksync/server.go

```diff
@@ -0,0 +1,262 @@
+package blocksync
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"time"
+
+	"go.opencensus.io/trace"
+	"golang.org/x/xerrors"
+
+	cborutil "github.com/filecoin-project/go-cbor-util"
+
+	"github.com/filecoin-project/lotus/chain/store"
+	"github.com/filecoin-project/lotus/chain/types"
+
+	"github.com/ipfs/go-cid"
+	inet "github.com/libp2p/go-libp2p-core/network"
+)
+
+
+// BlockSyncService is the component that services BlockSync requests from
+// peers.
+//
+// BlockSync is the basic chain synchronization protocol of Filecoin. BlockSync
+// is an RPC-oriented protocol, with a single operation to request blocks.
+//
+// A request contains a start anchor block (referred to with a CID), and a
+// amount of blocks requested beyond the anchor (including the anchor itself).
+//
+// A client can also pass options, encoded as a 64-bit bitfield. Lotus supports
+// two options at the moment:
+//
+//  - include block contents
+//  - include block messages
+//
+// The response will include a status code, an optional message, and the
+// response payload in case of success. The payload is a slice of serialized
+// tipsets.
+// FIXME: Rename to just `Server` (will be done later, see note on `BlockSync`).
+type BlockSyncService struct {
+	cs *store.ChainStore
+}
+
+func NewBlockSyncService(cs *store.ChainStore) *BlockSyncService {
+	return &BlockSyncService{
+		cs: cs,
+	}
+}
+
+// Entry point of the service, handles `Request`s.
+func (server *BlockSyncService) HandleStream(stream inet.Stream) {
+	ctx, span := trace.StartSpan(context.Background(), "blocksync.HandleStream")
+	defer span.End()
+
+	defer stream.Close() //nolint:errcheck
+
+	var req Request
+	if err := cborutil.ReadCborRPC(bufio.NewReader(stream), &req); err != nil {
+		log.Warnf("failed to read block sync request: %s", err)
+		return
+	}
+	log.Infow("block sync request",
+		"start", req.Head, "len", req.Length)
+
+	resp, err := server.processRequest(ctx, &req)
+	if err != nil {
+		log.Warn("failed to process request: ", err)
+		return
+	}
+
+	_ = stream.SetDeadline(time.Now().Add(WRITE_RES_DEADLINE))
```

Comment:
need to clear the deadline too

Reply:
addressed in https://github.com/filecoin-project/lotus/pull/2715/commits/65512193ccf97e6fd9b22c9953086a1ea3ad967a